### DISCUSSION TITLE
startChannelScan and StartReceive for a CSMA/CA protocol

### DISCUSSION THREAD

**Discussion by Jaimi5** [jgromes/RadioLib#543] [Q&A] [answered]
I am trying to improve the collisions that are happening in my code, I am using the SX127x module. And I have found that scanning the channel before sending the packet is a good idea [CSMA/CA](https://en.wikipedia.org/wiki/Carrier-sense_multiple_access_with_collision_avoidance).

I have seen that there were some closed issues that mention this type of issue and I have not found an excellent solution for my code, but here is what I have now.
 1. Add setDioAction(receiveFunction)
 2.  startReceive()
 3. When want to send a packet I have a random delay and then StartChannelScan. I wait for the Dio0Action trigger which will trigger my receiveFunction. Inside here I have this lines of code:
 ```
uint16_t flags = radio->getIRQFlags();
if ((flags & RADIOLIB_SX127X_CLEAR_IRQ_FLAG_CAD_DETECTED) == RADIOLIB_SX127X_CLEAR_IRQ_FLAG_CAD_DETECTED) {
    addFlagDetected = 1;
    continue;
else if ((flags & RADIOLIB_SX127X_CLEAR_IRQ_FLAG_CAD_DONE) == RADIOLIB_SX127X_CLEAR_IRQ_FLAG_CAD_DONE) {
    //PreambleDoneWithoutDetection
    continue;
}

//Here readData...
```
4. Before sending, clearDIO0Action(), transmit.
5. After sending, setDIO0Action(receiveFunction) and startReceive() again.

But when I receive a packet and a CAD is detected, the DIO0 is not activated again by the startReceive().
I have encountered that after I received a preamble, and then I read the packet with readData, there are some times that this data is broken and the CRC does not work.

How could I receive this CAD detected and receive the packet at the same time?
I had seen some suggestions in #67. Having the startReceive that triggers DIO0 when a preamble is detected and wait for all the packet to be received to read all could be possible?

Thanks for your time! Hope this is not a dumb question.

**Comment by jgromes** [comment]
I have not tried this combination, so it's possible there could be some bug in the library. 

> How could I receive this CAD detected and receive the packet at the same time?

If you need to do both, why not just set the module to receive forever? Although the SX127x suggest CAD before Rx could be possible.

> Having the startReceive that triggers DIO0 when a preamble is detected and wait for all the packet to be received

Fully received packet will laos trigger DIO0, so it will be tricky for you to figure out which event triggered it.

**Comment by Jaimi5** [comment]
I tried some code and I have some results that previously get wrong.

When an interrupt of the startChannelScan is set to RADIOLIB_SX127X_CLEAR_IRQ_FLAG_CAD_DETECTED, I can read immediately the packet completely.
But I don't know why, but when I read the packet and that is smaller than the previous one, I will get the previous packet inside the FIFO. (previous packet received or sent).
But if the packet is bigger, I will perfectly receive the packet.

And when I read it the packetLenght, the RSSI and the SNR are always wrong, it gets the values of the previous packet and not the actual.

Here are some logs:

----- Good but bad packetLength, SNR and RSSI
01:15:03.164 > W: Receiving LoRa packet: Size: 12 bytes RSSI: -19 SNR: 10
01:15:03.171 > Packet received -- Size: 12 Src: 0x63AC Dst: 0xDD14 Id: 20 Type: 1010001 Via: 0xDD14 Seq_Id: 13 Num: 1
01:15:15.368 > W: RandomDelay 1920 ms
01:15:17.289 > W: Starting scan
01:15:17.290 > W: Preamble detected
01:15:17.296 > W: Receiving LoRa packet: Size: 12 bytes RSSI: -19 SNR: 10
01:15:17.301 > Packet received -- Size: 25 Src: 0x96A0 Dst: 0xFFFF Id: 2 Type: 100 

----- Good but bad packetLength, SNR and RSSI
01:11:15.333 > W: Receiving LoRa packet: Size: 12 bytes RSSI: -49 SNR: 9
01:11:15.333 > Packet received -- Size: 12 Src: 0x6D4C Dst: 0x8C20 Id: 7 Type: 1000 Via: 0x8C20 Seq_Id: 0 Num: 0
01:11:18.568 > W: Starting scan
01:11:18.584 > W: Preamble detected
01:11:18.584 > W: Receiving LoRa packet: Size: 12 bytes RSSI: -49 SNR: 9
01:11:18.584 > Packet received -- Size: 222 Src: 0x8C20 Dst: 0x6D4C Id: 2 Type: 10001 Via: 0x6D4C Seq_Id: 0 Num: 1

----- Get the previous received one
01:15:18.490 > W: Receiving LoRa packet: Size: 22 bytes RSSI: -36 SNR: 10
01:15:18.490 > Packet received -- Size: 22 Src: 0x8C20 Dst: 0xFFFF Id: 8 Type: 100 
01:15:18.567 > W: Starting scan
01:15:18.581 > W: Preamble detected
01:15:18.581 > W: Receiving LoRa packet: Size: 22 bytes RSSI: -36 SNR: 10
01:15:18.581 > Packet received -- Size: 22 Src: 0x8C20 Dst: 0xFFFF Id: 8 Type: 100 

----- Get the previous sent one, (Packet Length, RSSI and SNR from the previous received packet)
01:10:24.142 > W: Receiving LoRa packet: Size: 13 bytes RSSI: -10 SNR: 3
01:10:24.142 > Packet received -- Size: 13 Src: 0x63AC Dst: 0xFFFF Id: 0 Type: 100 
01:10:24.142 > W: RandomDelay 1440 ms
01:10:24.142 > W: Starting scan
01:10:24.142 > W: Preamble done
01:10:24.142 > Packet send -- Size: 7 Src: 0x9234 Dst: 0xFFFF Id: 0 Type: 100 
01:10:33.366 > W: RandomDelay 480 ms
01:10:33.846 > W: Starting scan
01:10:33.848 > W: Preamble detected
01:10:33.854 > W: Receiving LoRa packet: Size: 13 bytes RSSI: -10 SNR: 3
01:10:33.860 > Packet received -- Size: 7 Src: 0x9234 Dst: 0xFFFF Id: 0 Type: 100 


> If you need to do both, why not just set the module to receive forever? Although the SX127x suggest CAD before Rx could be possible.

Is there a way that I am not seeing? Is the startReceive(RXCONTINUOUS) what you mean?


Moreover, I don't think this is the intention of the StartChannelScan, but I don't know what could I do to fix that.

**Reply by Jaimi5** [reply]
> And when I read it the packetLenght, the RSSI and the SNR are always wrong, it gets the values of the previous packet and not the actual.

After reading the SX127x documentation I see why this does not work -> 

> In continuous mode status information are available only for the last packet received, i.e. the corresponding registers should be read before the next RxDone arrives.

I think the problem here is that when CADDone, the library does not change to a RXSingle or RXContinuous mode in the RX init, and then it does not trigger the RXDone. And doing startReceive() when the CAD_Detected, it will put the radio in standby for a moment and it loses the packet.
![CADMode](https://user-images.githubusercontent.com/33554367/180411685-a4d97c88-febf-4b4d-a774-35659965e72e.PNG)

Is there a possibility to fix it or do a work arround to improve the startChannelScan()?
I am sorry if I am a little dumb... I am a noob with hardware and I am sorry if I am misunderstanding all the time.

**Reply by jgromes** [reply]
There's no silly questions here, quite the contrary - everything is on point.

> And doing startReceive() when the CAD_Detected, it will put the radio in standby for a moment and it loses the packet.

In that case, it could be argued that the call to standby shouldn't be there. In fact, `startReceive` is only being called from the blocking `receive` method, which already puts the module to standby. So removing it shouldn't break anything in the library. Have you tried what happens if you do that and then call startReceive after CAD is detected?

**Comment by Jaimi5** [comment, marked as answer]
> Have you tried what happens if you do that and then call startReceive after CAD is detected?

Yes, and it did not work... It does not trigger the RXDone.
In addition, the Dio1Action is not triggering when CAD finished or detected.

If you need me to try other combinations tell me!

**Reply by Jaimi5** [reply]
Hi @jgromes, is there something I could help or test to try fix this? Thanks!

**Reply by jgromes** [reply]
@Jaimi5 thanks for reminding me! I added an example implementing what you're trying to accomplish: https://github.com/jgromes/RadioLib/blob/master/examples/SX127x/SX127x_Channel_Activity_Detection_Receive/SX127x_Channel_Activity_Detection_Receive.ino

Check it out ;)

**Reply by Jaimi5** [reply]
@jgromes It seams to work now with your example! Thanks you very much for your work!
